db/bbolt.go

```diff
@@ -64,80 +64,68 @@ func (bb *BBoltHandler) CreateInitialSchema() error {
 	})
 }
 
-func (bb *BBoltHandler) InsertBlock(block *types.Block) error {
-	bb.logger.Info("Inserting block to DB", zap.Uint64("block_height", block.BlockHeight))
-
-	// Store mapping number -> block
-	err := bb.db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket([]byte(blocksBucket))
-		key := bb.itob(block.BlockHeight)
-		blockBytes, err := json.Marshal(block)
-		if err != nil {
-			return err
-		}
-		return b.Put(key, blockBytes)
-	})
-	if err != nil {
-		bb.logger.Error("Error inserting block", zap.Error(err))
-		return err
+func (bb *BBoltHandler) InsertBlocks(blocks []*types.Block) error {
+	if len(blocks) == 0 {
+		return nil
 	}
 
-	// Store mapping hash -> number
-	err = bb.db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket([]byte(blockHeightsBucket))
-		return b.Put([]byte(block.BlockHash), bb.itob(block.BlockHeight))
-	})
-	if err != nil {
-		bb.logger.Error("Error inserting block", zap.Error(err))
-		return err
-	}
+	bb.logger.Info("Batch inserting blocks to DB", zap.Int("count", len(blocks)))
+
+	// Single transaction for all operations
+	return bb.db.Update(func(tx *bolt.Tx) error {
```

Comment:
this PR is already too large to review. it's probably better to keep those refactoring stuff in a base PR and have the core logic changes stacked on top of it.

for example, a base PR can merge all the `bb.db.Update` calls inside this function. and other less important stuffs such as adding the config validation logic (that can be its own PR actually)

no need to split anything from this PR tho. just bring up the discussion point

Reply:
got it, yeah will try to do this going forward. was prioritising speed but appreciate it prob made code review more difficult!